src/docmanager/index.ts

```diff
@@ -0,0 +1,958 @@
+      let widgets: Widget[] = this._widgets[id] || [];
+      for (let w of widgets) {
+        w.close();
+      }
+    }
+  }
+
+  /**
+   * Close all of the open documents.
+   */
+  closeAll(): void {
+    for (let id in this._widgets) {
+      for (let w of this._widgets[id]) {
+        w.close();
+      }
+    }
+  }
+
+  /**
+   * Create a context or reuse an existing one.
+   */
+  private _createContext(path: string, model: IDocumentModel, widgetName: string, contents: IContentsModel): string {
+    let mFactoryEx = this._getModelFactoryEx(widgetName);
+    let id = this._contextManager.findContext(path, mFactoryEx.name);
+    if (id) {
+      return id;
+    } else {
+      return this._contextManager.createNew(path, model, mFactoryEx, contents);
+    }
+  }
+
+  /**
+   * Create a widget from a context and attach it to the parent.
+   */
+  private _createWidget(contextId: string, widgetName: string, parent: Widget, kernel?: IKernelId): void {
+    let wFactoryEx = this._getWidgetFactoryEx(widgetName);
+    if (!(contextId in this._widgets)) {
+      this._widgets[contextId] = [];
+    }
+    this._widgets[contextId].push(parent);
+    let context = this._contextManager.getContext(contextId);
+    let model = this._contextManager.getModel(contextId);
+    // Create the child widget using the factory.
+    let child = wFactoryEx.factory.createNew(model, context, kernel);
+    this._attachChild(parent, child);
+    Private.nameProperty.set(parent, widgetName);
+    Private.contextProperty.set(parent, contextId);
+    installMessageFilter(parent, this);
+  }
+
+  /**
+   * Attach a child widget to a parent container.
+   */
+  private _attachChild(parent: Widget, child: Widget) {
+    parent.layout = new PanelLayout();
+    parent.title.closable = true;
+    parent.title.text = child.title.text;
+    parent.title.icon = child.title.icon;
+    parent.title.className = child.title.className;
+    // Mirror the parent title based on the child.
+    child.title.changed.connect(() => {
+      child.parent.title.text = child.title.text;
+      child.parent.title.icon = child.title.icon;
+      child.parent.title.className = child.title.className;
+    });
+    // Add the child widget to the parent widget.
+    (parent.layout as PanelLayout).addChild(child);
+  }
+
+  /**
+   * Ask the user whether to close an unsaved file.
+   */
+  private _maybeClose(widget: Widget, dirty: boolean): Promise<boolean> {
+    if (!dirty) {
+      return Promise.resolve(true);
+    }
+    let host = widget.isAttached ? widget.node : document.body;
+    return showDialog({
+      title: 'Close without saving?',
+      body: `File "${widget.title.text}" has unsaved changes, close without saving?`,
+      host
+    }).then(value => {
+      if (value && value.text === 'OK') {
+        return true;
+      }
+      return false;
+    });
+  }
+
+  /**
+   * Clean up the data associated with a widget.
+   */
+  private _cleanupWidget(widget: Widget): void {
```

Comment:
It feels like this function has unnecessary complications. Furthermore, it's not called when, for example, a widget with a clean model is closed. Why don't we have the parent widget (the one this doc manager creates) have a dispose() function that removes it from the appropriate `this._widgets` list. Then calling the widget's dispose function automatically cleans it up from the doc manager.